File: src/practise.rb
require_relative("import")

class Practise
	@beatmap	= nil
	@time		= (Struct.new(:start, :end)).new
	@beginning	= nil
	@rng		= (Struct.new(:time, :position)).new
	@hardrock	= false
	@skip	= false

	class << self
		attr_accessor :beatmap, :output, :time, :beginning, :rng, :hardrock, :skip
	end

	def self.add_beginning(time, amount)
		if Practise.beginning.nil?
			Practise.beginning = Array.new
		end
		beg = (Struct.new(:time, :amount)).new
		beg[:time] = time
		beg[:amount] = amount
		Practise.beginning.push(beg)
	end

	def self.split_time(str)
		str_split = str.split(",", 0)
		if str_split.length != 2
			puts("Error: Time [#{t}] requires the start time and end time")
			exit(1)
		end
		for i in 0..str_split.length-1 do
			str_i = str_split[i].split(":", 0)
			if str_i.length == 3
				time_i = str_i.map(&:to_i)
				str_split[i] = (time_i[0] * 60000) + (time_i[1] * 1000) + time_i[2]
			else
				str_split[i] = str_split[i].to_i
			end
		end
		return str_split
	end
end

def practise_rename(map)
	type = !Practise.rng[:time].nil? || !Practise.rng[:position].nil? ? (Practise.hardrock ? "hrd" : "rng") : "nmd"
	version = "#{map[:metadata][:version]} - #{type} #{Practise.time[:start]}-#{Practise.time[:end]}"
	$output_global = LIBOSU.fopen("./#{map[:metadata][:artist]} - #{map[:metadata][:title]} (#{map[:metadata][:creator]}) [#{version}].osu", "w")
	if $output_global.null?
		puts("Error: Unable to create output file in current directory.")
		exit(1)
	end
	LIBOSU.ofb_metadata_setfromstring(map[:metadata], "Version:#{version}")
	temp, temp_num = map[:hit_objects], map[:num_ho]
	map[:hit_objects], map[:num_ho] = nil, 0
	LIBOSU.of_beatmap_tofile($output_global, map)
	map[:hit_objects], map[:num_ho] = temp, temp_num
end

def practise_rng(map)
	if Practise.rng[:time].nil? || Practise.rng[:position].nil?
		return
	end

	rng = LIBOSU::LegacyRandom.new
	LIBOSU.ou_legacyrandom_init(rng, LIBOSU.ooc_processor_RNGSEED)

	last_position = FFI::MemoryPointer.new(:pointer)
	last_start_time = FFI::MemoryPointer.new(:double)

	rng_num = 0

	map[:num_ho].times do | i |
		hit_object = LIBOSU::HitObject.new(map[:hit_objects].to_ptr + (i * LIBOSU::HitObject.size))
		if hit_object[:time] >= Practise.time[:start]
			next
		end

		object = LIBOSU::CatchHitObject.new
		case hit_object[:type]
		when :circle, :nc_circle
			LIBOSU.ooc_fruit_init(object, hit_object)
		when :slider, :nc_slider
			LIBOSU.ooc_juicestream_initwslidertp(object, map[:difficulty], map[:timing_points], map[:num_tp], hit_object);
			LIBOSU.ooc_juicestream_createnestedjuice(object);
		when :spinner, :nc_spinner
			LIBOSU.ooc_bananashower_init(object, hit_object)
			LIBOSU.ooc_bananashower_createnestedbananas(object)
		end

		old_rng = LIBOSU::LegacyRandom.new
		old_rng[:w] = rng[:w]
		old_rng[:x] = rng[:x]
		old_rng[:y] = rng[:y]
		old_rng[:z] = rng[:z]

		LIBOSU.ooc_processor_applypositionoffsetrng(object, 1, rng, Practise.hardrock, last_position, last_start_time)

		while !(old_rng[:w] == rng[:w] && old_rng[:x] == rng[:x] && old_rng[:y] == rng[:y] && old_rng[:z] == rng[:z])
			LIBOSU.ou_legacyrandom_nextuint(old_rng)
			rng_num += 1
		end
	end

	rng_div = (rng_num / (4 * 2)).to_i
	rng_mod = rng_num % (4 * 2)

	rng_div.times do | i |
		bs_ho = LIBOSU::HitObject.new
		bs_ho[:x] = 256
		bs_ho[:y] = 192
		bs_ho[:time] = Practise.rng[:time]
		bs_ho[:type] = :nc_spinner
		bs_ho[:hit_sound] = 0
		bs_ho[:ho][:spinner][:end_time] = Practise.rng[:time] + 1
		output = FFI::MemoryPointer.new(:pointer)
		LIBOSU.ofb_hitobject_tostring(output, bs_ho);
		LIBOSU.fprintf($output_global, output.read_pointer.read_string)
	end

        $beg_num = 0
	rng_mod.times do | i |
		if $js_faster_length.nil? || $js_faster_length == 1
			$js_faster_length = 1
		end
		js_ho = LIBOSU::HitObject.new
		js_ho[:x] = Practise.rng[:position]
		js_ho[:y] = 384
		js_ho[:time] = Practise.rng[:time]
		js_ho[:type] = :nc_slider
		js_ho[:hit_sound] = 0
		js_ho[:ho][:slider][:curve_type] = :slidertype_linear
		js_ho[:ho][:slider][:num_curve] = 1
		js_ho[:ho][:slider][:curves] = LIBOSU::HOSliderCurve.new
		js_ho[:ho][:slider][:curves][:x] = Practise.rng[:position]
		js_ho[:ho][:slider][:curves][:y] = 0
		js_ho[:ho][:slider][:slides] = 1
		js_ho[:ho][:slider][:length] = $js_faster_length

		loop do
			# For some reason, while looping, this gets GCed, so we have to reinitialise it ever time
			js_ho[:ho][:slider][:curves] = LIBOSU::HOSliderCurve.new
			js_ho[:ho][:slider][:curves][:x] = Practise.rng[:position]
			js_ho[:ho][:slider][:curves][:y] = 0

			js_obj = LIBOSU::CatchHitObject.new
			LIBOSU.ooc_juicestream_initwslidertp(js_obj, map[:difficulty], map[:timing_points], map[:num_tp], js_ho)
			LIBOSU.ooc_juicestream_createnestedjuice(js_obj)
			js_new = LIBOSU::JuiceStream.new(js_obj[:cho][:js])
			if js_new[:num_nested] == 3
				output = FFI::MemoryPointer.new(:pointer)
				LIBOSU.ofb_hitobject_tostring(output, js_ho);
				LIBOSU.fprintf($output_global, output.read_pointer.read_string)
                                $beg_num += 1
				break
			end
			$js_faster_length += 1
			js_ho[:ho][:slider][:length] = $js_faster_length
		end
	end
end

def practise_beginning(map)
	if Practise.beginning.nil?
		return
	end
	Practise.beginning.each do | i |
		if i[:amount] == 0
			ho = 0
			(map[:num_ho] - $beg_num).times do | j |
				map_ho = LIBOSU::HitObject.new(map[:hit_objects].to_ptr + (j * LIBOSU::HitObject.size))
				if map_ho[:time] >= Practise.time[:start] || ho == 200
					break
				else
					if map_ho[:type] == :slider || map_ho[:type] == :nc_slider
						js_cho = LIBOSU::CatchHitObject.new
						LIBOSU.ooc_juicestream_initwslidertp(js_cho, map[:difficulty], map[:timing_points], map[:num_tp], map_ho)
						LIBOSU.ooc_juicestream_createnestedjuice(js_cho)
						js = LIBOSU::JuiceStream.new(js_cho[:cho][:js])
						js[:num_nested].times do | k |
							object = LIBOSU::CatchHitObject.new(js[:nested].to_ptr + (k * LIBOSU::CatchHitObject.size))
							if ho == 200
								break
							elsif object[:type] == :catchhitobject_fruit || object[:type] == :catchhitobject_droplet
								ho += 1
							end
						end
					elsif map_ho[:type] == :circle || map_ho[:type] == :nc_circle
						ho += 1
					end
				end
			end

			hit_object = LIBOSU::HitObject.new
			hit_object[:x] = 256
			hit_object[:y] = 192
			hit_object[:time] = i[:time]
			hit_object[:type] = :circle
			hit_object[:hit_sound] = 0

			(ho-2).times do
				output = FFI::MemoryPointer.new(:pointer)
				LIBOSU.ofb_hitobject_tostring(output, hit_object);
				LIBOSU.fprintf($output_global, output.read_pointer.read_string)
			end

			2.times do
				hit_object[:type] = :nc_circle
				hit_object[:time] += 1
				output = FFI::MemoryPointer.new(:pointer)
				LIBOSU.ofb_hitobject_tostring(output, hit_object);
				LIBOSU.fprintf($output_global, output.read_pointer.read_string)
			end
		else
			i[:amount].times do
				hit_object = LIBOSU::HitObject.new
				hit_object[:x] = 256
				hit_object[:y] = 192
				hit_object[:time] = i[:time]
				hit_object[:type] = :nc_circle
				hit_object[:hit_sound] = 0
				output = FFI::MemoryPointer.new(:pointer)
				LIBOSU.ofb_hitobject_tostring(output, hit_object);
				LIBOSU.fprintf($output_global, output.read_pointer.read_string)
			end
		end
	end
end

def practise_time(map)
	map[:num_ho].times do | i |
		hit_object = LIBOSU::HitObject.new(map[:hit_objects].to_ptr + (i * LIBOSU::HitObject.size))
		if hit_object[:time] < Practise.time[:start] || hit_object[:time] > Practise.time[:end]
			next
		end

		output = FFI::MemoryPointer.new(:pointer)
		LIBOSU.ofb_hitobject_tostring(output, hit_object);
		LIBOSU.fprintf($output_global, output.read_pointer.read_string)
	end
end

def practise_main
	# Create the map to store all the data in
	map = LIBOSU::Beatmap.new
	LIBOSU.of_beatmap_init(map)
	if Practise.skip
		map[:timing_points] = LIBOSU.realloc(map[:timing_points], 1 * LIBOSU::TimingPoint.size)
		map[:num_tp] = 1
		LIBOSU.ofb_timingpoint_addfromstring(map[:timing_points], "-1,6,1,0,0,100,1,0")
	end
	LIBOSU.of_beatmap_set(map, Practise.beatmap)

	# Rename the map to tell the player what section of the map they are playing
	practise_rename(map)

	# Populate map with the `beginning` argument
	practise_beginning(map)
	
	# Process rng and populate map based on the rng
	practise_rng(map)

	# Record only the requested sections of the map
	practise_time(map)

	# Close output file
	LIBOSU.fclose($output_global)
end

File: src/practise.c
#include "practise.h"

#include <osu.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

Practise practise = {
	.beatmap = NULL,

	.output = NULL,

	.time = {
		.start = 0,
		.end = 0
	},

	.beginning = NULL,
	.beginning_num = 0,
	
	.rng = {
		.time = 0,
		.position = 0
	},
	.hardrock = false
};

void practise_rename(Beatmap map) {
	int time_start = ou_comparing_size(practise.time.start);
	int time_end = ou_comparing_size(practise.time.end);
	int buffer_num = strlen("Version:") + 1 + 1 + time_start + 1 + time_end + 1;
	char *buffer = calloc(buffer_num, sizeof(*buffer));
	snprintf(buffer, buffer_num, "Version: %c%d-%d", practise.rng.time || practise.rng.position ? (practise.hardrock ? 'h' : 'r') : 'n', practise.time.start, practise.time.end);
	free(map.metadata->version);
	ofb_metadata_setfromstring(map.metadata, buffer);
	free(buffer);

	HitObject *temp = map.hit_objects;
	unsigned int temp_num = map.num_ho;
	map.hit_objects = NULL;
	map.num_ho = 0;
	of_beatmap_tofile(practise.output, map);
	map.hit_objects = temp;
	map.num_ho = temp_num;
}

void practise_beginning(Beatmap map) {
	for (int i = 0; i < practise.beginning_num; i++) {
		for (int j = 0; j < (practise.beginning + i)->amount; j++) {
			HitObject object = {
				.x = 256,
				.y = 192,
				.time = (practise.beginning + i)->time,
				.type = nc_circle,
				.hit_sound = 0,
				.hit_sample = {0}
			};
			char *output = NULL;
			ofb_hitobject_tostring(&output, object);
			fprintf(practise.output, "%s", output);
			free(output);
		}
	}
}

void practise_rng(Beatmap map) {
	LegacyRandom rng = {0};
	ou_legacyrandom_init(&rng, ooc_processor_RNGSEED);

	float *last_position = NULL;
	double last_start_time = 0;

	unsigned int rng_num = 0;

	for (int i = 0; i < map.num_ho; i++) {
		if ((map.hit_objects + i)->time >= practise.time.start) {
			break;
		}
		CatchHitObject *object = calloc(1, sizeof(*object));
		switch ((map.hit_objects + i)->type) {
			case circle:
			case nc_circle:
				ooc_fruit_init(object, (map.hit_objects + i));
				break;

			case slider:
			case nc_slider:
				ooc_juicestream_initwslidertp(object, *map.difficulty, map.timing_points, map.num_tp, (map.hit_objects + i));
				ooc_juicestream_createnestedjuice(object);
				break;

			case spinner:
			case nc_spinner:
				ooc_bananashower_init(object, (map.hit_objects + i));
				ooc_bananashower_createnestedbananas(object);
				break;
		}
		LegacyRandom old_rng = rng;
		ooc_processor_applypositionoffsetrng(object, 1, &rng, practise.hardrock, &last_position, &last_start_time);
		while (!(old_rng.w == rng.w
			&& old_rng.x == rng.x
			&& old_rng.y == rng.y
			&& old_rng.z == rng.z)) {
				ou_legacyrandom_nextuint(&old_rng);
				rng_num++;
		}
		ooc_hitobject_freebulk(object, 1);
	}

	free(last_position);

	unsigned int rng_div = rng_num / (4 * 2);
	unsigned int rng_mod = rng_num % (4 * 2);

	for (int i = 0; i < rng_div; i++) {
		HitObject bs_ho = {
			.x = 256,
			.y = 192,
			.time = practise.rng.time,
			.type = nc_spinner,
			.hit_sound = 0,
			.ho.spinner.end_time = practise.rng.time + 1,
			.hit_sample = {0}
		};
		char *output = NULL;
		ofb_hitobject_tostring(&output, bs_ho);
		fprintf(practise.output, "%s", output);
		free(output);
	}

	for (int i = 0; i < rng_mod; i++) {
		static unsigned int js_faster_length = 1;
		HitObject js_ho = {
			.x = practise.rng.position,
			.y = 384,
			.time = practise.rng.time,
			.type = nc_slider,
			.hit_sound = 0,
			.ho.slider = {
				.curve_type = slidertype_linear,
				.curves = NULL,
				.num_curve = 1,
				.slides = 1,
				.length = js_faster_length
			},
			.hit_sample = {0}
		};
		js_ho.ho.slider.curves = calloc(js_ho.ho.slider.num_curve, sizeof(*js_ho.ho.slider.curves));
		(js_ho.ho.slider.curves + js_ho.ho.slider.num_curve - 1)->x = practise.rng.position;
		(js_ho.ho.slider.curves + js_ho.ho.slider.num_curve - 1)->y = 0;
		while (true) {
			CatchHitObject js_obj = {0};
			ooc_juicestream_initwslidertp(&js_obj, *map.difficulty, map.timing_points, map.num_tp, &js_ho);
			ooc_juicestream_createnestedjuice(&js_obj);
			if (js_obj.cho.js->num_nested == 3) {
				char *output = NULL;
				ofb_hitobject_tostring(&output, js_ho);
				fprintf(practise.output, "%s", output);
				free(output);
				oos_hitobject_free(js_ho);
				ooc_hitobject_free(js_obj);
				break;
			}
			ooc_hitobject_free(js_obj);
			js_faster_length++;
			js_ho.ho.slider.length = js_faster_length;
		}
	}
}

void practise_time(Beatmap map) {
	for (int i = 0; i < map.num_ho; i++) {
		if ((map.hit_objects + i)->time < practise.time.start || (map.hit_objects + i)->time > practise.time.end) {
			continue;
		}
		char *output = NULL;
		ofb_hitobject_tostring(&output, *(map.hit_objects + i));
		fprintf(practise.output, "%s", output);
		free(output);
	}
}

void practise_main() {
	// Create the map to store all the data
	Beatmap map = {0};
	of_beatmap_init(&map);
	of_beatmap_set(&map, practise.beatmap);

	// Rename the map to tell the player what section of the map they are playing
	practise_rename(map);

	// Populate map with the `beginning` argument
	practise_beginning(map);

	// Process rng and populate map based on the rng
	practise_rng(map);

	// Record only the requested sections of the map
	practise_time(map);

	of_beatmap_free(map);
}